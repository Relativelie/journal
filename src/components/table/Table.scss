th {
    position: relative;
}

.filteringBtn {
    position: absolute;
    width: 100%;
    height: 100%;
}