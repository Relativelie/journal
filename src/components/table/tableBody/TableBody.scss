@import "../../../variables";

.tableBody__item {
    vertical-align: center;
    padding: 15px 12px 14px 11px;
    border-spacing: 0px;
    border: 1px solid #E3E6EC;
    @include textStyle($fontSize: 13px, $fontColor: $textFontColor)
}

.tableBody__item_id {
    text-align: center;
}

.tableBody__item div {
    height: 50px;
    overflow: auto;
}
