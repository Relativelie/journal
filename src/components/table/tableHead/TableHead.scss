@import "../../../variables";

.tableHead {
    background-color: #474955;

    &.id {
        width: 110px
    }

    &.title {
        width: 529px;
    }

    &.body {
        width: 433px;
    }
}
th {
    position: relative;
}

.tableHead__filteringBtn {
    position: absolute;
    width: 100%;
    height: 100%;
}

.tableHead__content {
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 31px;
}

.tableHead__content__par {
    @include textStyle($fontSize: 14px, $fontColor: $headTextColor)
}