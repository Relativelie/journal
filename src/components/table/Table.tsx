import 'regenerator-runtime/runtime';
import { useEffect } from 'react';
import { useActions } from '../../hooks/useActions';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import filterImage from '../../assets/images/filterImage.png';

import './Table.scss';

export const Table = () => {
    const { tableContent, tableContentCurrent } = useTypedSelector(
        (tableState) => tableState.tableReducer,
    );
    const { getResult } = useTypedSelector(
        (requestState) => requestState.sendRequestReducer,
    );
    const { isFiltered, filteredList, activeFilter } = useTypedSelector(
        (filterState) => filterState.filterReducer,
    );
    const {
        sendGet, saveTableContent, turnOffFilter, turnOnFilter, changeTableContent,
    } = useActions();

    useEffect(() => {
        const url = 'https://jsonplaceholder.typicode.com/posts';
        const headers = {
            Accept: 'application/json',
            'Content-Type': 'application/json',
        };
        sendGet(url, headers);
    }, []);

    useEffect(() => {
        changeTableContent(getResult);
        saveTableContent(getResult);
    }, [getResult]);

    useEffect(() => {
        if (isFiltered) changeTableContent(filteredList);
        else changeTableContent(tableContent);
    }, [activeFilter]);

    const filtering = (event: EventTarget) => {
        const elem = event as HTMLElement;
        if (activeFilter === elem.dataset.filterby) {
            turnOffFilter();
        } else if (elem.dataset.filterby !== undefined) {
            turnOnFilter(elem.dataset.filterby, tableContent);
        }
    };

    return (
        <main>
            <table>
                <tbody>
                    <tr>
                        <th>
                            <div
                                className="filteringBtn"
                                aria-label="filter values"
                                onClick={(e) => filtering(e.target)}
                                onKeyUp={(e) => filtering(e.target)}
                                data-filterby="id"
                                role="button"
                                tabIndex={0}
                            />
                            <p>ID</p>
                            <img src={filterImage} alt="filter" />
                        </th>
                        <th>
                            <div
                                className="filteringBtn"
                                aria-label="filter values"
                                onClick={(e) => filtering(e.target)}
                                onKeyUp={(e) => filtering(e.target)}
                                data-filterby="title"
                                role="button"
                                tabIndex={0}
                            />
                            <p>Заголовок</p>
                            <img src={filterImage} alt="filter" />
                        </th>
                        <th>
                            <div
                                className="filteringBtn"
                                aria-label="filter values"
                                onClick={(e) => filtering(e.target)}
                                onKeyUp={(e) => filtering(e.target)}
                                data-filterby="body"
                                role="button"
                                tabIndex={0}
                            />
                            <p>Описание</p>
                            <img src={filterImage} alt="filter" />
                        </th>
                    </tr>
                    {tableContentCurrent.map((item) => {
                        const { id, title, body } = item;
                        return (
                            <tr key={`tableItem-${id}`}>
                                <td>{id}</td>
                                <td>{title}</td>
                                <td>{body}</td>
                            </tr>
                        );
                    })}
                </tbody>
            </table>
        </main>
    );
};
