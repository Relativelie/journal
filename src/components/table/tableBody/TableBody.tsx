import { FC } from 'react';
import './TableBody.scss';

type Props = {
    paginatedContent: any[]
};

export const TableBody: FC<Props> = ({ paginatedContent }) => {
    return (
        <tbody>
            {paginatedContent.map((item) => {
                const { id, title, body } = item;
                return (
                    <tr
                        className="tableBody"
                        key={`tableItem-${id}`}
                    >
                        <td className="tableBody__item tableBody__item_id">{id}</td>
                        <td className="tableBody__item">{title}</td>
                        <td className="tableBody__item">
                            <div>{body}</div>
                        </td>
                    </tr>
                );
            })}
        </tbody>
    );
};
