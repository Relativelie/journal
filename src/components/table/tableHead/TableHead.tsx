import { FC } from 'react';
import filterImage from '../../../assets/images/filterImage.png';
import { FilterObj } from '../../../types/filterTypes';
import './TableHead.scss';

type Props = {
  filtering: Function;
  allFilters: FilterObj[];
  activeFilter: string;
};

export const TableHead: FC<Props> = ({
    filtering,
    allFilters,
    activeFilter,
}) => {
    return (
        <thead>
            <tr>
                {allFilters.map((item) => {
                    const { id, attributeName, headName } = item;
                    let imgClassName = '';
                    if (attributeName === activeFilter) {
                        imgClassName = 'tableHead__content_activeFilter';
                    }
                    return (
                        <th
                            key={`tableHeadItem-${id}`}
                            className={`tableHead ${attributeName}`}
                        >
                            <div
                                className="tableHead__filteringBtn"
                                aria-label="filter values"
                                onClick={(e) => filtering(e.target)}
                                onKeyUp={(e) => filtering(e.target)}
                                data-filterby={attributeName}
                                data-testid="table head"
                                role="button"
                                tabIndex={0}
                            />
                            <div className="tableHead__content">
                                <p className="tableHead__content__par">{headName}</p>
                                <img className={imgClassName} src={filterImage} alt="filter" />
                            </div>
                        </th>
                    );
                })}
            </tr>
        </thead>
    );
};
