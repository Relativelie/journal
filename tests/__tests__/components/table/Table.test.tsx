import { Table } from '../../../../src/components/table/Table';
import { TableBody } from '../../../../src/components/table/tableBody/TableBody';
import { TableHead } from '../../../../src/components/table/tableHead/TableHead';
import { render } from '../../../testsSetup/test-utils';
import { data } from '../../../testValues/data';

describe('Table component', () => {
    test('snapshot - table, tableHead, tableBody component', () => {
        const filtering: Function = jest.fn();
        const tree = render(
            <Table>
                <TableHead
                    filtering={filtering}
                    allFilters={[{ id: 1, attributeName: 'id', headName: 'ID' }]}
                    activeFilter="id"
                />
                <TableBody paginatedContent={data} />
            </Table>,
        );
        expect(tree).toMatchSnapshot();
    });
});
