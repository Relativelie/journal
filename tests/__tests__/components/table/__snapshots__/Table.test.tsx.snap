// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component snapshot - table, tableHead, tableBody component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <table
          class="table"
        >
          <thead>
            <tr>
              <th
                class="tableHead id"
              >
                <div
                  aria-label="filter values"
                  class="tableHead__filteringBtn"
                  data-filterby="id"
                  data-testid="table head"
                  role="button"
                  tabindex="0"
                />
                <div
                  class="tableHead__content"
                >
                  <p
                    class="tableHead__content__par"
                  >
                    ID
                  </p>
                  <img
                    alt="filter"
                    src="test-file-stub"
                  />
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="tableBody"
            >
              <td
                class="tableBody__item tableBody__item_id"
              >
                1
              </td>
              <td
                class="tableBody__item"
              >
                sunt aut facere repellat provident occaecati excepturi optio reprehenderit
              </td>
              <td
                class="tableBody__item"
              >
                <div>
                  quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
                </div>
              </td>
            </tr>
            <tr
              class="tableBody"
            >
              <td
                class="tableBody__item tableBody__item_id"
              >
                2
              </td>
              <td
                class="tableBody__item"
              >
                qui est esse
              </td>
              <td
                class="tableBody__item"
              >
                <div>
                  est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
                </div>
              </td>
            </tr>
            <tr
              class="tableBody"
            >
              <td
                class="tableBody__item tableBody__item_id"
              >
                3
              </td>
              <td
                class="tableBody__item"
              >
                ea molestias quasi exercitationem repellat qui ipsa sit aut
              </td>
              <td
                class="tableBody__item"
              >
                <div>
                  et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </main>
    </div>
  </body>,
  "container": <div>
    <main>
      <table
        class="table"
      >
        <thead>
          <tr>
            <th
              class="tableHead id"
            >
              <div
                aria-label="filter values"
                class="tableHead__filteringBtn"
                data-filterby="id"
                data-testid="table head"
                role="button"
                tabindex="0"
              />
              <div
                class="tableHead__content"
              >
                <p
                  class="tableHead__content__par"
                >
                  ID
                </p>
                <img
                  alt="filter"
                  src="test-file-stub"
                />
              </div>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="tableBody"
          >
            <td
              class="tableBody__item tableBody__item_id"
            >
              1
            </td>
            <td
              class="tableBody__item"
            >
              sunt aut facere repellat provident occaecati excepturi optio reprehenderit
            </td>
            <td
              class="tableBody__item"
            >
              <div>
                quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
              </div>
            </td>
          </tr>
          <tr
            class="tableBody"
          >
            <td
              class="tableBody__item tableBody__item_id"
            >
              2
            </td>
            <td
              class="tableBody__item"
            >
              qui est esse
            </td>
            <td
              class="tableBody__item"
            >
              <div>
                est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
              </div>
            </td>
          </tr>
          <tr
            class="tableBody"
          >
            <td
              class="tableBody__item tableBody__item_id"
            >
              3
            </td>
            <td
              class="tableBody__item"
            >
              ea molestias quasi exercitationem repellat qui ipsa sit aut
            </td>
            <td
              class="tableBody__item"
            >
              <div>
                et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
